### ISSUE TITLE
Create an application process for using services or join groups

### ISSUE THREAD

**Issue by bum2** [FreedomCoop/valuenetwork#107]
The idea is to do the same we did for membership process, but for using legal services (and inside ocp).
- One form page to gather needed special data
- A new page for admins to receive all applications and moderate use cases (also do new ocp groups to manage that)
- A new page for members with the same structure as the membership-request page: the service form content (some fields editable) and a comments thread related the desired service.
- The menu item link to this page when there's an application process for the user: "service activation" (once accepted the menu item can be renamed to 'service feedback').
- The same link can be added in the applicant project agent page.

If someone has a better idea on how to solve this, please talk.

**Comment by bum2** [comment]
I've been thinking a more general way to make a moderated process of accepting new members or participants in some of the ocp groups (context agents). If we can do this, we'll have a simple way to make the application, review, commenting and acceptation of participants in any group, without doing a form (model, db table) for each group. It's basically using some of the stuff we did for FC membership but with a more general approach, so it seems easy and fast to develop:
- Add an option in the context agents configuration (new field?) to choose group joining style:
    - moderated: will need a form to fill, a comments page and a requests table for project managers.
    - autojoin: perhaps can be defined the accepted auto associations (maybe only participant or supporter) for each group. 
- Add an option in the context agents config (new field?) to choose the visibility:
    - 'private': visible only for group sub-agents (group members, participants, etc) or admins
    - 'only FCmembers': visible for any Freedom Coop member.
    - 'public': visible for anybody.
- Make a general 'project join request' model, with a subset of the fields used in FC Membership (without 'shares', 'type' nor some other fields), and a new 'ForeignKey' field to relate the choosed context-agent or group.
- Make a 'join group form' form page with two versions, one for the workapp (logged users) and one for the outside (non logged).

If a group choose 'moderated' joining and 'public' visibility (like in the Faircoin2 CVN process), the group name shows in a list of 'public' groups in a generic external form for non logged users. So the user goes to the public form, choose the desired group to join and fills in the personal data, description (or 'reasons', 'motivation'), references, etc.

- Build a 'project-join-requests-table' for project managers in workapp: same as membership-request table but in the 'members_agent' page, opened from 'your projects' page, when the user is manager of that project. The new agent association dropdown of the table will list only related new 'candidate' users of that group.
- The project manager cannot yet create the new users and agents (an admin will be needed to do them nowadays) but he can link the new user with their join-request table row. Also this project managers can see the link to the comments page for every candidate and receive email notifications and/or new tasks when a new request is made or a new user's comment is placed.
- Build a 'join-project-process' personal page for these new group candidates (like the actual comments page) and put a link to it in the menu (if the user have this ongoing process) like 'Join Requests'.
- This menu item can be a dropdown with many 'join request comments' page links if the user is applying for more than one 'moderated' projects. The menu item link can be shown using the group name (eg. "Join FairCoin2 CVN group") inside this menu dropdown ('Join Requests'). After the candidate user is accepted (when project managers change that user-group relation, from 'candidate' to 'member' or 'participant'), the comments page for the join process can go to the 'username' menu, with another name: "Projects Feedback" and subitems like "{group name} Feedback".

If a group chooses 'only FCmembers', the external form don't show them. Inside OCP we can do a 'Join Projects' page with a list of groups 'open for FreedomCoop members' that links to the internal version of the join form for each project.

Many other things can be done with this model, but now is proposed to do only the most urgent 'use case' for the CVN process, and later we can develop more the other 'use cases'.

**Comment by bum2** [comment]
With this general proposed approach we can't include special fields for every project, so only common users data is asked, and any needed special users data must be asked and answered using the comments thread. This way all requests of all groups are using the same db table and django model.

**Comment by mario567** [comment]
Good, seems to be a practical solution. Beside the CVN process would be good to test this way of groups access for legal tools services like the sporadic invoices.  Your mentioned service form with editable fields (+comments) would be very usable i guess as we can make a pre-selection of the services we can offer to the specific user case. As i understand it right, the membership process to "service goups" would be still manual, right?!

**Comment by enricduran** [comment]
How complex it would become if each form can have specific fields or just a different form?

It seems very important this personalization, for being really useful

**Comment by bhaugen** [comment]
Django packages has several apps for creating custom forms and fields:
https://github.com/stephenmcd/django-forms-builder
https://github.com/philomat/django-form-designer
https://github.com/MarkusH/django-dynamic-forms
https://github.com/barseghyanartur/django-fobi
https://github.com/mishbahr/djangocms-forms
https://github.com/django-dataforms/django-dataforms
https://github.com/django-dataforms/django-dataforms
https://github.com/roalddevries/django-dynamicforms
https://github.com/zbyte64/django-fieldmaker
(for example).

**Comment by bum2** [comment]
The personalization and different forms seems more difficult to achieve with the actual software. If we can let the specifics of each project to the comments thread, then we can make this general version of a moderation access system and form, usable for any project, with the tools we have developed so far.
I'm thinking on general ways to let projects define and manage specific form fields in their join-request process, but the development needed is very more complex. So a fast solution is this simplyfied general version of the existent 'membership-request' model (that will also take some hot days of developing). In a future release can be added the feature of adding special fields in a project join form, integrating some new plugin into ocp.

**Comment by XaviP** [comment]
I think for the maintenance easier of the code, this feature can be included as a new app, and the fields needed for each context agent configuration (enumerated by @bum2) can be included in the model of the new app.

**Comment by bhaugen** [comment]
@bum2 was asking for some help in doing this. If we want a new app, I could start helping by creating the new app in a new branch.

**Comment by bum2** [comment]
Ok, let's do it!
Perhaps Bob can focus in the agent model part, extending the model with the fields:

> Add an option in the context agents configuration (new field?) to choose group joining style:
>-    moderated: will need a form to fill, a comments page and a requests table for project managers.
>-    autojoin: perhaps can be defined the accepted auto associations (maybe only participant or supporter) for each group.

> Add an option in the context agents config (new field?) to choose the visibility:
>-    'private': visible only for group sub-agents (group members, participants, etc) or admins
>-    'only FCmembers': visible for any Freedom Coop member.
>-    'public': visible for anybody.

I can try to make a general version of the membership-request model (called join-request) in the same new app and branch... what can be the app name? 'join_a_project'? or simply 'project'? ... 
By the way i can't code tomorrow... i will focus on this sunday night or monday.

Also had a look on the list of plugins for flexible forms, and i think the https://github.com/barseghyanartur/django-fobi and the https://github.com/stephenmcd/django-forms-builder looks good (what are your choices?), but i'm afraid we can not integrate this kind of thing for next week... so i'm focusing in a fast solution.

**Comment by bhaugen** [comment]
Here's the first draft of the Project model:
```
JOINING_STYLE_CHOICES = (
    ('moderated', _('moderated')),
    ('autojoin', _('autojoin')),
)

VISIBILITY_CHOICES = (
    ('private', _('private')),
    ('FCmembers', _('only FC members')),
    ('public', _('public')),
)

class Project(models.Model):
    agent = models.OneToOneField(EconomicAgent,
        verbose_name=_('agent'), related_name='project')
    joining_style = models.CharField(_('joining style'),
        max_length=12, choices=JOINING_STYLE_CHOICES,
        default="moderated")
    visibility = models.CharField(_('visibility'),
        max_length=12, choices=VISIBILITY_CHOICES,
        default="private")
```

Please suggest improvements. If you got an agent, you can say agent.project and get the related instance of this class (if one exists). If you got an instance of this class, you can say project.agent, and get the related agent.

@bum2 you mentioned a field for a flexible form key, too, but I can't add that until I can reference a flexible form app model.

**Comment by bum2** [comment]
Thanks Bob! The model looks good to start working on!
From the list of plugins you provided for 'flexible forms', which plugin seems better to you? in a fast lookup i saw two interesting ones but not tested nor studied in deep, to integrate in OCP... will be good to have an easy way to open such a config tool for project managers, or at least for coop admins, but i think they're all only for superusers (the definition part).
Even if we don't use such a tool now, will be good to have the project access Form as a separate model, linked with this 'form_id' field, but i'm not sure if that's for the 'project' model because is only needed if they choose 'moderated' access. Many projects will not have a join-form. So maybe we need a ModeratedProject as a extension of this Project model? what you think about? or let the 'form_id' field empty when not needed?
If is only this field (using a forms plugin) is not much, but if we want all the info to customize a generic form then they are many fields, only used when project access is moderated, so not very good to live in the Project model...

**Comment by bhaugen** [comment]
> From the list of plugins you provided for 'flexible forms', which plugin seems better to you?

I'll check out the two you listed above and report what I learn back here.

**Comment by bhaugen** [comment]
I would not be afraid of either https://github.com/barseghyanartur/django-fobi or the https://github.com/stephenmcd/django-forms-builder

Fobi is packaged and documented better (including a nice online demo), but is still working on 1.9 migration. Forms-builder claims to work on 1.9. 

Fobi requires a lot more configuration, but thus is more flexible.

Forms-builder relies on Django admin, while it looks like Fobi offers their own form creation pages.

Forms-builder offers a list of form entries (in Django admin). I'm not seeing one in Fobi, but it looks like it would be easy to create one, and I don't think we want that in Django admin anyway.

So I think I would start by  trying Fobi in another new branch and unless it fails in some way, stick with it.

**Comment by bum2** [comment]
Marvelous Bob! it looks a good one...

**Comment by XaviP** [comment]
A doubt: are you going to setup a new model for the project join-requests? Or you will use the membership request model for projects too?

**Comment by bum2** [comment]
I was proposing to have a 'form_id' field in the new Project model, and link there the custom form created for the project.
- If we can use the Fobi plugin for flexible forms, the form_id FK will be the Fobi form id.
- If we can't use the flexible forms yet (if something fails), then we need a new 'JoinRequest' model as a generic form for joining a project (then the form_id FK is the JoinRequest.pk), and also a way to customize the strings for some fields. 

So, is much better to go with the Fobi way, even if initially managed by admins or superusers.
This form_id field will be Null, if the Project.join_style is not 'moderated'.
Many of the fields are common as new ocp users and agents, so not coming from Fobi or JoinRequest form models.

**Comment by bhaugen** [comment]
@bum2 - I got fobi installed in a local branch. Up and running. Looks good. Do you want to take it from here?

**Comment by bum2** [comment]
Ok, if you put the fobi branch in github i'll take it and continue from there.
The thing i'm doing is the 'create new project' button in 'your-projects' page, but still not ready, due a problem applying the migrations at 'project' branch.
Is pending to add an admin's feature to link a context agent with a project (in coop admin app), defining the project new fields. Do you want to do that part?

**Comment by bhaugen** [comment]
> problem applying the migrations at 'project' branch.

Did you see my response in Telegram?

> Is pending to add an admin's feature to link a context agent with a project (in coop admin app), defining the project new fields. Do you want to do that part?

Shouldn't that be done by the project manager in your-projects? If you want me to do that, I'll do it when you push your update to the project branch.

**Comment by bhaugen** [comment]
>  if you put the fobi branch in github i'll take it and continue from there.

Pushed. You will need to do
```
pip install django-fobi
```
and then do runserver and then go to http://127.0.0.1:8000/fobi/
See the fobi readme and doc for more info:
https://github.com/barseghyanartur/django-fobi
http://django-fobi.readthedocs.io/en/latest/#

**Comment by bhaugen** [comment]
Ran into a bug in fobi, reported here: https://github.com/barseghyanartur/django-fobi/issues/12

https://github.com/barseghyanartur apparently closed that issue before. We'll see how he responds to problems.

**Comment by bhaugen** [comment]
A bit more fobi testing: that bug is not very important. It only happens in django admin. I can get the form data entries other ways.

**Comment by bhaugen** [comment]
@bum2 if and when you get into fobi, I can give you some tips.

**Comment by bum2** [comment]
Ok, the Fobi thing is working and somewhat integrated into the work app also. The workflow that is working now is:
- Project coordinator chooses 'moderated' joining style (appears the Form Slug field) in the project editor page.
- He lists the custom project fields (questions) to add to a basic form with common fields (JoinRequest model) in the comments thread with admins.
- Admins must do the custom part of the form in the fobi pages (not self-managed by coordinators yet) and put the form title that becames the form slug  (url part)
- Admins or managers set this 'fobi_slug' in the Project editor.
- If the Project visibility is set to 'public' the form just appears for non-logged visitors in the url joinaproject/<fobi_slug>
- Visitors can fill the form and it appears a note in the project page, and a new row in the table at the new 'Join Requests' project subpage, and also a task (create agent and user) is created for the project.
- In that table is now the agents dropdown to assign a newly created, but i want to change this:
I want to put a 'Make Candidate User' button to do all that with a click: create sub agent 'candidate' + create a ocp user (the password will need to be automatic?)
After being candidate, the button disapears and in that cell it shows the link to the comments thread on the join-request feedback page (still doing this), and a button to 'Make Active Participant' and erase the table row.

**Comment by bum2** [comment]
The fobi plugin is not in the repo (we installed via "pip install django-fobi"). There is no touch to the fobi code, but there's a ForeignKey to one of its models (SavedFormDataEntry at fobi/contrib/plugins/form_handlers/db_store/models.py) and i don't know the implications of this FK... Can we keep the fobi plugin in the env or in the system of the server or it is better to put the fobi app code in the repo?
When i did makemigrations it adds also a migration in the fobi folder, just for the FK to work... should we create and keep this migration file in the fobi folder of the env or system of any machine?

**Comment by bhaugen** [comment]
> (the password will need to be automatic?)

You might be able to use:
https://docs.djangoproject.com/en/1.8/topics/auth/customizing/#django.contrib.auth.models.BaseUserManager.make_random_password
but you would need to communicate the password to the new user somehow.

Alternatively, we could implement an invite-user flow, customizing the flow that is somewhat built into the account app that we are already using in OCP. The big difference is that an OCP user needs an agent as well as a user, and that agent will already have been created by the project manager and associated with the project.

**Comment by bhaugen** [comment]
> Can we keep the fobi plugin in the env or in the system of the server or it is better to put the fobi app code in the repo?

It's better to keep fobi as an external requirement and put it in requirements.txt.

**Comment by bhaugen** [comment]
> When i did makemigrations it adds also a migration in the fobi folder, just for the FK to work

Are you sure? Where can I find this migration? Unless you are changing fobi itself, adding a foreign key to a fobi model in an OCP model should not change fobi. If it does, that's seems like a django migrations problem that I would like to discuss both with fobi and the django core team.

**Comment by bum2** [comment]
let's see if that fobi migration is needed or not in another install... can you pull the fobi branch and try for yourself the workflow without any migration to fobi folder? perhaps is something local of my machine or env, or can also be a mistake...
The fobi folder migraton file is not in the git scope, so not in the repo.

**Comment by bhaugen** [comment]
Fobi comes with its own migrations. When you install it the first time, you need to do migrations to get the fobi tables into the OCP database. If that is all, it's not a problem. If you did something that created a migration inside fobi, that is a problem, because it will not be available in test or prod ocp unless we fork fobi and use our fork.

We internalized the account app (rather than forking it) because the version we were using looked like it was no longer maintained. We probably should have forked it, but I was in a hurry and it was getting in the way so I got lazy.

**Comment by bhaugen** [comment]
Fobi in my site-packages has two migrations. How many do you see?

**Comment by bum2** [comment]
> You might be able to use:
> https://docs.djangoproject.com/en/1.8/topics/auth/customizing/#django.contrib.auth.models.BaseUserManager.make_random_password
> but you would need to communicate the password to the new user somehow.

> Alternatively, we could implement an invite-user flow, customizing the flow that is somewhat built into the account app that we are already using in OCP. The big difference is that an OCP user needs an agent as well as a user, and that agent will already have been created by the project manager and associated with the project.

I'm thinking to produce the pass in a easy known way and send an email to the candidates with the credentials and saying 'please change your password asap'. If we produce it random we need to store it somewhere and it also has to be communicated and recomended to change anyway.
Can you work on this part Bob? the button to create agent+user from the JoinRequest data automatically... the new user becames 'candidate' participant of the project and receives an email with credentials to log in and start the commenting moderation part.
After this button is done, the next button to code is the one for 'Accept as Active Participant' (appearing once the first button is clicked) just to make the relation to 'active' and hide the row in the join_requests page table. All this happens in the table cell is used now for link an agent, in the work/agent/x/join-requests page. Also a needed thing is the associations page for projects, so they can manage them more freely.
I can keep working in the creation of the comments page for join requests if you can do this buttons and behaviours

**Comment by XaviP** [comment]
> it shows the link to the comments thread on the join-request feedback page

- The join-request comments template can be cloned from:
https://github.com/FreedomCoop/valuenetwork/blob/master/work/templates/work/membership_request_with_comments.html#L140

- This template prepares the variables in the view (that can be cloned too):
https://github.com/FreedomCoop/valuenetwork/blob/master/work/views.py#L1532

- I can see the new model JoinRequest in work/models.py (branch fobi) If we want email notification as in membership_request, the notifications are ready to be added with this new model in:
https://github.com/FreedomCoop/valuenetwork/blob/master/work/__init__.py#L32
But maybe notifications are a second stage, when we already have all this stuff tested and working in production.

**Comment by bhaugen** [comment]
> Can you work on this part Bob? the button to create agent+user from the JoinRequest data automatically... the new user becames 'candidate' participant of the project and receives an email with credentials to log in and start the commenting moderation part.

Ok. I think I got a couple of other things to do, will ask on the one-week sprint thread about priorities.

**Comment by bum2** [comment]
The buttons are already in the new master merged with fobi branch. Only the automation of create agent and user button is still missing. The password will need to be autocreated, but not random (we talk that by chat), then sended to user by email, and then will be good to enforce the user to change his password once logged-in for the first time (now we can just add a text about this in the home)...

**Comment by XaviP** [comment]
This feature is included in new v0.9.4
Closing. Please reopen if needed.
